include/rocksdb/utilities/replayer.h

```diff
@@ -61,13 +65,40 @@ class Replayer {
   // trace;
   // Status::NotSupported() if the operation is not supported;
   // Otherwise, return the corresponding error status.
-  virtual Status Execute(const std::unique_ptr<TraceRecord>& record) = 0;
-  virtual Status Execute(std::unique_ptr<TraceRecord>&& record) = 0;
+  //
+  // The actual operation execution status and result(s) will be saved in
+  // result. For example, a GetQueryTraceRecord will have its DB::Get() status
+  // and the returned value save in a SingleValueTraceExecutionResult.
+  virtual Status Execute(const std::unique_ptr<TraceRecord>& record,
+                         std::unique_ptr<TraceRecordResult>* result) = 0;
+  inline Status Execute(const std::unique_ptr<TraceRecord>& record) {
+    return Execute(record, nullptr);
+  }
+
+  virtual Status Execute(std::unique_ptr<TraceRecord>&& record,
+                         std::unique_ptr<TraceRecordResult>* result) = 0;
+  inline Status Execute(std::unique_ptr<TraceRecord>&& record) {
+    return Execute(std::move(record), nullptr);
+  }
```

Comment:
I feel like these API overloads are proliferating unnecessarily:

- This PR changes which overload is pure virtual, so it's not compatible (which is fine - we haven't released this code yet so can feel free to make incompatible changes). But if we're not aiming for compatibility, why keep the single argument `Execute()`s?
- I can't think of what scenario we'd need overloads for taking unique_ptr by const lvalue reference, and by rvalue reference:
  - If we take ownership of record we only need the rvalue reference interface (std::unique_ptr is not copyable).
  - If we don't take ownership of record we only need the lvalue reference interface.

My guess is `virtual Status Execute(const std::unique_ptr<TraceRecord>& record, std::unique_ptr<TraceRecordResult>* result) = 0;` is good enough.

Reply:
Done